components/fxa-client/src/lib.rs

```diff
@@ -243,10 +254,29 @@ impl FirefoxAccount {
         }
     }
 
-    fn get_refresh_token(&self) -> Result<&str> {
+    fn get_refresh_token(&mut self) -> Result<&str> {
         match self.state.refresh_token {
             Some(ref token_info) => Ok(&token_info.token),
-            None => Err(ErrorKind::NoRefreshToken.into()),
+            None => {
+                log::info!("No refresh_token, attempting to recover via migration state");
+                // if we have no refresh token, check if this is the case of a failed migration
+                if self.state.in_flight_migration.is_some() {
+                    // if there's a pending migration state, then try to provision a refresh_token
+                    // by exchanging the session token for the refresh token
```

Comment:
nit: this is not an "exchange" because you keep the sessionToken as well, you are using the sessionToken to grant the refresh token.